libraries/api/src/middleware/logging.js

```diff
@@ -0,0 +1,41 @@
+/**
+ * Log an API request on completion, including information determined
+ * by the `remoteAuthentication` middleware, if present.
+ */
+const logRequest = ({builder, entry, monitor}) => {
+  const apiMethod = `${builder.serviceName}.${builder.apiVersion}.${entry.name}`;
+  return (req, res, next) => {
+    let sent = false;
+    const start = process.hrtime();
+    const send = async () => {
+      try {
+        // Avoid sending twice
+        if (sent) {
+          return;
+        }
+        sent = true;
+
+        const d = process.hrtime(start);
+
+        monitor.log.apiMethod({
+          apiMethod,
+          public: req.public,
+          hasAuthed: req.hasAuthed,
+          resource: req.originalUrl,
+          method: req.method,
+          clientId: req.hasAuthed ? await req.clientId() : '',
+          satisfyingScopes: req.satisfyingScopes ? req.satisfyingScopes : [],
+          statusCode: res.statusCode,
+          duration: d[0] * 1000 + d[1] / 1000000,
+        });
+      } catch (err) {
+        monitor.reportError(err);
```

Comment:
Should we consider moving the `monitor.log.apiMethod` into a `finally` of this try/catch? I don't want there to be many situations where it doesn't end up getting logged.

Reply:
The only other bit in the function is a conditional and `process.hrtime`, though.. we'd add more code to handle the case where `duration` wasn't set, and include that in the `finally` block.

Reply:
Maybe we should just remove the try/catch in that case.